torchtune/modules/transforms/_transforms.py

```diff
@@ -0,0 +1,164 @@
+            >>> intervals = transform._get_image_attention_intervals(tokens)
+            >>> print(intervals)
+            [(0, 7), (1, 7), (7, 12)]
+        """
+        end = len(tokens)
+        vision_token_locations = [
+            i for i, token in enumerate(tokens) if token == self.image_token_id
+        ]
+        # Return empty list if there are no images
+        if len(vision_token_locations) == 0:
+            return []
+        # If there is only one image, it will attend to subsequent text until end
+        if len(vision_token_locations) == 1:
+            return [[vision_token_locations[0], end]]
+
+        # Construct intervals from previous image token to next image token
+        vision_masks = [
+            [tok_idx_prev, tok_idx_next]
+            # Offset by one to get consecutive indices
+            for tok_idx_prev, tok_idx_next in zip(
+                vision_token_locations[:-1], vision_token_locations[1:]
+            )
+        ]
+        # Last image will attend to subsequent text until end
+        vision_masks.append([vision_token_locations[-1], end])
+
+        # If there are consecutive vision tokens, they should all attend to the
+        # same subsequent text
+        last_mask_end = vision_masks[-1][1]
+        for vision_mask in vision_masks[::-1]:
+            if vision_mask[0] == vision_mask[1] - 1:
+                vision_mask[1] = last_mask_end
+            last_mask_end = vision_mask[1]
+        return vision_masks
+
+    def __call__(self, *, tokens: List[int], images: List[torch.Tensor], **kwargs):
+        """
+        Generates the vision cross-attention mask for the given sample based on
+        the image token locations interleaved in the text sequence.
+
+        Args:
+            tokens (List[int]): List of token IDs in the text sequence. Number of
+                image token IDs in the sequence must match the number of images.
+            images (List[torch.Tensor]): List of image Tensors post-tiling of shape
+                (n_tiles, c, h, w) each.
+            **kwargs (Dict[str, Any]): all other keys within the sample that will
+                not be altered by this transform.
+
+        Returns:
+            Dict[str, Any]: updated sample with the following keys:
+                - encoder_mask (List[torch.Tensor]): masks of shape (text_seq_len, image_seq_len)
+                - tokens (List[int]): original tokens
+                - images (List[torch.Tensor]): original images
+        """
+
+        # One sample can have multiple images - verify the number of image tokens
+        # is the same
+        n_img = len(images)
+        intervals = self._get_image_attention_intervals(tokens)
+        assert len(intervals) == n_img
+
+        # Create mask for each individual image based on its number of tokens,
+        # which can vary based on number of tiles. The masks are padded and concatenated
```

Comment:
noob question: what does "and concatenated" mean here?

Reply:
if we have multiple images in a sample, their masks need to be concatenated after padding upto max tiles per image

Reply:
made the comment more clear